### DISCUSSION TITLE
How about Pitchbend?  [solved]

### DISCUSSION THREAD

**Discussion by ErichHeinemann** [marcel-licence/esp32_basic_synth#12] [Ideas]
Hi Marcel, 
today I started to look into that code.
I would like to combine it with a crappy sampleplayer to mix these sounds.

In Samplerbox(.org) they have a working pitchbend.
How could we add a global pitchbend? Any idea to fake a wrong "samplefrequency" ? Or we change the osc->samplepos ?
+/-12 Tones are mostly the normal range.

Velocity and a global Volume is missing too.
I will check Your code in the next days, perhaps I am able to understand it a bit better. :)
.. or do You think You could manage this in Your code?

Best Regards
Erich

**Comment by marcel-licence** [comment]
Hi Erich,

I've some code I could merge into the project from a current one which does allow pitchbend. 
At the moment the frequency is controlled by this statement:
`osc->samplePos += osc->addVal;`

For pitchbend the addVal value can be calculated by changing the midi parameter. Also modulation would be possible. In that case the parameter should be refreshed continously. This may reduce the max supported polyphony.
This could be achieved by a multiplier.

Multiply by 0.5 goes one octave down, multply by 2 one octave up.
For semitones the multiplier is calculated by 2^(semtone/12).
For cents you could use 2^(cents/50/12). In that case 50 cents is one half semitone.

I can merge the code soon if you like.

Best regards,
Marcel

**Reply by ErichHeinemann** [reply]
Hi Marcel, that sounds cool!!

**Reply by ErichHeinemann** [reply]
Hi Marcel, - I have tried it by my own but if I multiple it like:
osc->samplePos += (osc->addVal) *0.5 ;
I will only get a distored sound .... it seems to bee more tricky :)

**Comment by ErichHeinemann** [comment]
Works :)
One Octave up and one down - super!

Added one global variable in esp32_basic_synth.ino
// Pitchbend
float globalBend = 1.0;


Added Pitchbendcontroller in my modified midi_interface.ino: ( based on Arduino MIDI Library)

void handlePitchBend( byte channel, int bend ){
  
  // Pitchbend is not a CC! and it is 14Bit!
  
  Serial.printf("Pitchbend %X %d\n", channel, bend);    
  // The Bend-Value goes from -8192 to 8191 
  // Normalize the Bend to a Value from 0 to 1 to use it in the sketch for effects; 
#ifdef FILTERCHANNEL
  if( channel == FILTERCHANNEL ){
#endif 
  globalBend = 1;
  if( bend > 0){
    // linear bend up
    globalBend = (float) 1.0f + (1.0f *bend) / 8191;
  } 

  if( bend < 0){
    // linear Bedn down: 0.5 - 1 = -8192 : 0
    globalBend = (float) 1.0f +  (1.0f * bend) / ( 8192*2 );
  } 

 
  Serial.printf("GlobalBend %X %3f\n", channel, globalBend );
 
  // Or someting if pitch is negativ and something else if the bend is positiv ... 
#ifdef FILTERCHANNEL
  }
#endif

}




Then changed lines in easySynth.ino:


    /*
     * oscillator processing -> mix to voice
     */
    for (int i = 0; i < MAX_POLY_OSC; i++ ){
        oscillatorT *osc = &oscPlayer[i];
        { 
            osc->samplePos += (uint32_t) (globalBend * osc->addVal);            
            float sig = osc->waveForm[WAVEFORM_I(osc->samplePos)];
            osc->dest[0] += osc->pan_l * sig;
            osc->dest[1] += osc->pan_r * sig;
        }
    }

    /*
     * voice processing
     */

**Reply by marcel-licence** [reply]
Perfect! Thanks for your feedback. 
---
I did forget in the stepper motor project there is also some pitch bend and modulation applied. 

If I will find some time I will update this project on sunday

**Reply by marcel-licence** [reply]
How is the max polphony affected in your case adding the multiplier (globalBend) ?

**Reply by marcel-licence** [reply]
I messed up my code today, but finally I got also the modulation working and merged it into the project. I didn't know that the execution time of "pow()" will vary in case of using an input unequal to zero.

**Reply by pc0rba** [reply]
Hi Marcel,
"edirol pcr-800 " isnt supports MIDI Key Pressure message- 0xA0?
Most electronic keyboards are designed so that when pressure is applied a small change in pitch occurs similar to the tremolo system on electric guitars.

**Reply by marcel-licence** [reply]
Hi,
you are right. There is some aftertouch support. By using it it feels like breaking the keys but a good and valid point. Maybe I could think about some modifiers for the sound and apply to pressure or other things.

**Reply by pc0rba** [reply]
trbl- noteon(off)/ADSR. may be used without attack,only Sustain.

**Comment by ErichHeinemann** [comment]
Hi Marcel,
it is not a battle. I was only thinking and testing how it could be done and finally, I was able to add Pitchbend yesterday. Today I did some kind of Pitch-Modulation too. - it is in my fork. I think the only difference is my preference for the Arduino MIDI Library. All other libraries are still the same like Yours. You really did a great job and every new project made public from You is some kind of Christmas-Present :)

**Reply by marcel-licence** [reply]
Hi Erich,

thank you very much! You are right. 
I also often like to try out some ideas. Sometimes you can get very nice effects by random. But I also got some weird effects very often. I tried some modulation some days ago and on low notes it did only some kind of jumping between some frequencies.
Sounds great I am happy to hear that. The next present is waiting :-D It will be based on one of your video comments.

**Comment by ErichHeinemann** [comment]
Hi Marcel,
I did not recognise any effect on it.
I dropped some lines of debug-messages which where specially in my code.

As my code is not the same as Yours, 


 oscillatorT *osc = &oscPlayer[i];
        {
            // portamento has to be calculated per Oscillator by incrementing from old osc->addVal to the curreent addVal.  
            osc->samplePos += (uint32_t) ( ( 1 + ( lfo_sig * lfo_matrix_to_pitch )) * globalBend * osc->addVal  );            

As my loo_sig and Leo_matrix_to_pitch are botch small positiv/negative floats, it is implemented as a multiplier.
Low notes are bended less from the numeric perspective and higher notes are bended more .. as I did ist as a multiplier.
As both values could be positiv or negative I had to add 1  see  ( 1 + ( flo… )

It works without pow.

In my code I had implemented some „matrix“ - variables to be able to connect the LFO as a Source to different destinations .. .probably Attack, Delay etc.
But not only a few are implemented.
I also tested with a high frequency for LFO and it sounded a bit as a crappy FM :)

My Version ist here:
https://github.com/ErichHeinemann/esp32_basic_synth/blob/main/easySynth.ino

Best Regards
Erich
